### PR TITLE
Simulation architecture refactor

### PR DESCRIPTION
**Pull request by dzeber**
This continues the refactoring in #36 of the simulation runner to an actual client-server model. The new design is outlined across a few new classes, but details still need to be filled in.

Summary of changes:

- `Client` class to encapsulate client data and manage weight updates individually
- `SGDModel` class to wrap the sklearn model. This lets model params get set once in the notebook where the simulation is run from and get shared around by cloning. We can also develop an interface to training which is better adapted to our needs, ie updating model weights on the fly and minibatch updating.
- Simulation runner redesign. Changes from `runner.py`:
  * Data loading will be handled in the notebook where the simulation is run from
  * Model and initial weights are initialized in the notebook
  * Params can be handled more simply and transparently using keyword args. This will also throw errors automatically if required params are missing.
  * Most of `run_fed_learn_sim` will get relegated to the calling notebook, ie. train/test split, weight initialization, specifying the param grid, validation, etc. (still in progress)
- Sketch of test cases

### CODE REVIEW

File: mozfldp/client.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        Resulting weights are submitted to the server.
+        """
+        self._model.set_weights(current_coef, current_intercept)
+
+        # TODO: split data into batches.
+
+        # TODO: update weights across batches and epochs.
+        for epoch in range(num_epochs):
+            for i in range(num_batches):
```
Comment (crankycoder) [inline mozfldp/client.py]: `num_batches` is undefined - maybe `batch_size`?

File: mozfldp/client.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                self._run_model_update_step(batch_features[i], batch_labels[i])
```
Comment (crankycoder) [inline mozfldp/client.py]: should `batch_features` and `batch_labels` be `self._features` and `self._labels`?

File: mozfldp/model.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from sklearn.linear_model import SGDClassifier
+from sklearn.base import clone
+from sklearn.utils.multiclass import unique_labels
+
```
Comment (crankycoder) [inline mozfldp/model.py:8]: missing `import numpy as np`

File: mozfldp/model.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        internally in computing an adaptive learning rate.
+        """
+        self.classifier.coef_ = coef
+        self.classifier.intercept_ = intercept
+
+    def get_weights(self):
+        """Return the current model weights as (coef, intercept)."""
+        return (self.classifier.coef_, self.classifier.intercept_)
+
+    def minibatch_update(self, X, y):
```
Comment (crankycoder) [inline mozfldp/model.py:71]: mismatch between `X` and `x` on line 82 ```for i in range(len(x)):```

File: mozfldp/model.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        mini_batch_intercept = np.zeros(init_weight[1].shape, dtype=np.float64, order="C")
+
+        for i in range(len(x)):
+            self.classifier.partial_fit(X[i],y[i])
+
+            coef = self.classifier.coef_
+            intercept = self.classifier.intercept_
+
+            # save the weight per sample
+            new_coefs = np.add(new_coefs, coef)
```
Comment (crankycoder) [inline mozfldp/model.py]: Should these be `mini_batch_coefs` and `mini_batch_intercept`?

File: mozfldp/simulation_runner.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        self._coefs = [coef_init]
+        self._intercepts = [intercept_init]
+        self._num_rounds_completed = 0
+
+        # Initialize the clients with their respective datasets.
+        self._clients = []
+        by_user_data = training_data.groupby("user_id")
+        for user_id, user_data in by_user_data:
+            feats, labs = _format_data_for_model(user_data, label_col, user_id_col)
+            self._clients.append(Client(user_id, feats, labs, self._model.get_clone()))
```
Comment (crankycoder) [inline mozfldp/simulation_runner.py]: Shouldn't this `get_clone` call require `true` to use a pretrained model?

**Review by crankycoder — COMMENTED** [review]
I think a short sprint to get some test cases around this refactoring will get things into shape for a merge.

File: mozfldp/model.py (hunk 4 above)
Comment (dzeber) [reply mozfldp/model.py:71]: This now applies to #43

File: mozfldp/model.py (hunk 5 above)
Comment (dzeber) [reply mozfldp/model.py]: This now applies to #43

File: mozfldp/model.py (hunk 3 above)
Comment (dzeber) [reply mozfldp/model.py:8]: This now applies to #43

File: mozfldp/simulation_runner.py (hunk 6 above)
Comment (dzeber) [reply mozfldp/simulation_runner.py]: I think you're right. I was originally thinking to initialize clients with an untrained model, but this should probably just pull in whatever state the supplied model is in. Ie. to start a fresh simulation, the caller should take care to pass in an untrained model.

File: mozfldp/client.py (hunk 2 above)
Comment (crankycoder) [reply mozfldp/client.py]: patched in 9d820b459fed259b7a2ec6362cc5a5b9d7f2475d

File: mozfldp/client.py (hunk 1 above)
Comment (crankycoder) [reply mozfldp/client.py]: patched in 9d820b459fed259b7a2ec6362cc5a5b9d7f2475d

**Comment by crankycoder** [comment]
@dzeber are we just waiting for some testcases to get this merged in?

**Comment by dzeber** [comment]
@crankycoder I filled out the implementation for the client update looping over batches (merged over your update as we were working on this concurrently). Also added some test cases. If you have a chance maybe give it a final look over, and I think this is ready to merge.

**Pull request merged**